doc/BALANCING.md

```diff
@@ -0,0 +1,389 @@
+  transaction.
+  
+* For the second option (`ManualFee`), the fee provided by the user will
+  be associated to the unbalanced transaction without adjustment.
+
+These choices ensure that fees are managed appropriately according to the needs
+of the transaction and the desired balance between cost and performance.
+
+### Collateral utxos
+
+Which utxos to pick from as collateral inputs.
+
+``` haskell
+data CollateralUtxos
+  = CollateralUtxosFromBalancingWallet -- default
+  | CollateralUtxosFromWallet Wallet
+  | CollateralUtxosFromSet (Set Api.TxOutRef) Wallet
+```
+
+In addition to the regular UTXOs consumed in a transaction, additional UTXOs
+must be provided to cover potential phase 2 validation failures. These UTXOs
+need to be sufficient to meet a specified total collateral requirement,
+typically 1.5 times the transaction fee based on protocol parameters. Any
+surplus can be returned to a designated wallet through an output known as return
+collateral. This setting determines which UTXOs the balancing mechanism should
+consider for inclusion in the transaction. Similar to
+[`BalancingUtxos`](#balancing-utxos), the final set of UTXOs included may not
+necessarily match the considered set, especially for collaterals, as protocol
+parameters also impose limits on the number of allowable collateral UTXOs
+(typically 3).
+
+Here are the options available:
+
+* `CollateralUtxosFromBalancingWallet`: Use UTXOs containing only value from the
+  balancing wallet. The return collateral will be directed back to the balancing
+  wallet. This option is synonymous with `CollateralUtxosFromWallet
+  (balancingWallet)`.
+* `CollateralUtxosFromWallet Wallet`: Use UTXOs containing only value from the
+  specified wallet. Transactions using these UTXOs will require the signing of
+  the wallet owner for validation. The return collateral will also be sent to
+  this same wallet.
+* `CollateralUtxosFromSet (Set Api.TxOutRef) Wallet`: Utilize UTXOs from the
+  provided set and direct return collaterals to the designated wallet. Note that
+  if any of these UTXOs belong to a script and are selected by the balancing
+  mechanism, validation will fail because only UTXOs controlled by public keys
+  are permissible as collaterals according to ledger rules. Exercise caution
+  with this option as it may involve losing additional information contained
+  within the chosen UTXO during the balancing process.
+
+If auto-balancing is disabled, the collateral calculation will still be based on
+the fee determined by the [`FeePolicy`](#fee-policy). Currently, users cannot
+specify the exact amount of collateral to be used in a transaction, unlike
+fees. This limitation is intentional to minimize user interaction with the
+collateral mechanism, which is rarely needed. Additionally, `cooked-validators`
+does not support issuing new transactions after a validation failure, making the
+collateral mechanism less relevant outside of transaction validation
+requirements.
```

Comment:
That's a very good point. I will try to actually test that and write down a sentence or two about it once I've ensured the mechanism works.

Reply:
Actually we cannot test that yet, because a failure just erases any kind of inner state of the mockchain and thus we cannot test properties on the resulting state. I think this would be a good time to make an issue about this, where we actually try to do something with validation errors instead of just spreading them. In the meantime, I will change the sentence to make it clear that for now we don't really do anything with collaterals.

Reply:
Here is the revised paragraph:
```
If auto-balancing is disabled, the collateral calculation will still be based on
the fee determined by the [`FeePolicy`](#fee-policy). Currently, users cannot
specify the exact amount of collateral to be used in a transaction, unlike
fees. This limitation is intentional to minimize user interaction with the
collateral mechanism, which is rarely needed. Additionally, `cooked-validators`
does not support issuing new transactions after a validation failure, making the
collateral mechanism irrelevant outside of transaction validation requirements.
However, future enhancements should allow validation failures to be treated as
acceptable, thereby increasing the usefulness of the collateral mechanism and
potentially introducing new skeleton options related to it.
```

Reply:
Great :+1: